Add loader enable/disable env vars

MarkY-LunarG opened the pull request:
Add new environment variables that will allow the loader to filter
layers and drivers in specific ways.
This control should give developers ways to more quickly narrow down
issues with a layer or driver.
Also, it should give CI environments a mechanism to selectively enable
only drivers and layers that are needed for testing.

Add tests to support validating the new changes.

Add documentation that describes the new changes.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 3568.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1446 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1446 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 10712.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1455 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1455 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 12130.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1456 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1456 passed.

charles-lunarg commented on docs/LoaderDebugging.md:
```diff
@@ -0,0 +1,312 @@
+but this will provide more layer-specific info as well such as:
+  * What layers are found
+  * Where they were found
+  * If they are implicit what environment variables can be used to disable them
```
```suggestion
  * If they are implicit, what environment variables can be used to disable them
```

charles-lunarg commented on loader/loader.c:
```diff
@@ -767,10 +768,241 @@ VkResult loader_add_to_dev_ext_list(const struct loader_instance *inst, struct l
+                                          struct loader_envvar_filter *filter_struct) {
+    memset(filter_struct, 0, sizeof(struct loader_envvar_filter));
+    char *env_var_value = loader_secure_getenv(env_var_name, inst);
+    if (NULL != env_var_value) {
+        if (strlen(env_var_value) > 0) {
```
Style wise, its better to do `if (NULL == env_var_value ) return;`, so the success path isn't as indented.

MarkY-LunarG replied:
👍🏽

charles-lunarg commented on loader/loader.c:
```diff
@@ -767,10 +768,241 @@ VkResult loader_add_to_dev_ext_list(const struct loader_instance *inst, struct l
+                }
+                loader_instance_heap_free(inst, parsing_string);
+            } else {
+                loader_log(inst, VULKAN_LOADER_ERROR_BIT, 0,
```
This should return OOM in this case.

MarkY-LunarG replied:
👍🏽

charles-lunarg commented on loader/loader.c:
```diff
@@ -767,10 +768,241 @@ VkResult loader_add_to_dev_ext_list(const struct loader_instance *inst, struct l
+                                                 struct loader_envvar_disable_layers_filter *disable_struct) {
+    memset(disable_struct, 0, sizeof(struct loader_envvar_disable_layers_filter));
+    char *env_var_value = loader_secure_getenv(VK_LAYERS_DISABLE_ENV_VAR, inst);
+    if (NULL != env_var_value) {
+        if (strlen(env_var_value) > 0) {
```
Same as above, this should early-out in the error cases.

MarkY-LunarG replied:
👍🏽

charles-lunarg commented on loader/loader.c:
```diff
@@ -767,10 +768,241 @@ VkResult loader_add_to_dev_ext_list(const struct loader_instance *inst, struct l
+            const size_t env_var_len = strlen(env_var_value);
+            // Allocate a separate string since strtok modifies the original string
+            char *parsing_string = loader_instance_heap_calloc(inst, env_var_len + 1, VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
+            if (NULL != parsing_string) {
```
This needs to return an OOM error in the failure path.

MarkY-LunarG replied:
👍🏽

charles-lunarg commented on loader/loader_common.h:
```diff
@@ -462,3 +462,29 @@ struct loader_msg_callback_map_entry {
+#define MAX_ADDITIONAL_FILTERS 16
+struct loader_envvar_filter {
+    uint32_t count;
+    struct loader_envvar_filter_value filters[VK_MAX_EXTENSION_NAME_SIZE];
```
Shouldn't this code use the `MAX_ADDITIONAL_FILTERS` value?

MarkY-LunarG replied:
👍🏽

charles-lunarg commented on loader/loader.c:
```diff
@@ -4129,49 +4468,144 @@ void loader_deactivate_layers(const struct loader_instance *instance, struct loa
 
+    if (NULL != vk_inst_layers) {
+        for (uint32_t cur_layer = 0; cur_layer < separator_count; ++cur_layer) {
+            if (NULL != vk_inst_layers[cur_layer]) {
```
Don't need to check for null here, as it'll happen inside the `free` call.

MarkY-LunarG replied:
👍🏽

charles-lunarg commented on loader/loader.c:
```diff
@@ -767,10 +768,241 @@ VkResult loader_add_to_dev_ext_list(const struct loader_instance *inst, struct l
+                               const struct loader_envvar_filter *filter_struct) {
+    bool ret_value = false;
+    const size_t name_len = strlen(name);
+    char *lower_name = loader_instance_heap_calloc(inst, name_len + 1, VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
```
This should report OOM

MarkY-LunarG replied:
👍🏽

charles-lunarg commented on loader/loader.c:
```diff
@@ -767,10 +768,241 @@ VkResult loader_add_to_dev_ext_list(const struct loader_instance *inst, struct l
+    bool ret_value = false;
+    const size_t name_len = strlen(name);
+    char *lower_name = loader_instance_heap_calloc(inst, name_len + 1, VK_SYSTEM_ALLOCATION_SCOPE_INSTANCE);
+    if (NULL != lower_name) {
```
Should early out here too.

MarkY-LunarG replied:
I switched this to a static size variable.  I don't want to have to worry about OOM here.

charles-lunarg commented on loader/loader.c:
```diff
@@ -3921,6 +4240,23 @@ void loader_scan_for_implicit_layers(struct loader_instance *inst, struct loader
+    // We'll need to save the dl handles so we can close them later
+    if (instance_layers->count > 0 && NULL != libs) {
+        *libs = loader_calloc(NULL, sizeof(loader_platform_dl_handle) * instance_layers->count, VK_SYSTEM_ALLOCATION_SCOPE_COMMAND);
+        if (*libs == NULL) {
```
This should set OOM.

MarkY-LunarG replied:
👍🏽

charles-lunarg commented on loader/loader.c:
```diff
@@ -4129,49 +4468,144 @@ void loader_deactivate_layers(const struct loader_instance *instance, struct loa
+        if ((NULL != disable_filter &&
+             (disable_filter->disable_all || disabled_by_type ||
+              check_name_matches_filter(inst, source_prop->info.layerName, &disable_filter->additional_filters))) &&
+            (NULL == enable_filter || !check_name_matches_filter(inst, source_prop->info.layerName, enable_filter))) {
+            loader_log(inst, VULKAN_LOADER_WARN_BIT | VULKAN_LOADER_LAYER_BIT, 0,
```
Do we need to check if the enable_filter here would enable the layer if we are going to do that later down below?

MarkY-LunarG replied:
Good point.  Verified removing this code has no effect on the tests.

charles-lunarg reviewed (changes requested):
Lots of small changes (mainly handling OOM and doing early-out for error paths).

I would like to have a go playing with this feature before putting it out there, just to make sure it works like I think it does.

hadess commented:
Looks like `VK_LOADER_DRIVER_SELECT=intel_icd*` would fix the vulkaninfo test case I submitted in https://github.com/KhronosGroup/Vulkan-Loader/issues/963, and work on both the host and Flatpak (given new enough version of the vulkan loader obviously).

I would however make sure in the docs to use "driver manifest filenames" consistently. "driver names" was used a couple of times even though the docs say "drivers don’t have a name".

MarkY-LunarG commented on docs/LoaderDebugging.md:
````diff
@@ -0,0 +1,312 @@
+```
+
+Then when the application selects the device to use, you will see the Vulkan
+device call chain reported in the following way (NOTE: additional spaces have
+been removed from the output for easier reading):
````
I'm not sure what you mean, do you mean you want the raw output combined with above or you want all the spaces put back?  It seems like a waste of GitHub viewing space to have the spaces since text might require scrolling right to see.  When we could just remove the spaces...

charles-lunarg replied:
I think I meant that there was a `...` in the example output, and I felt that there shouldn't be an ellipsis instead all the output should be given. 
However, as I see the text now it looks fine, so I'm not sure what I was objecting too.

MarkY-LunarG commented on docs/LoaderInterfaceArchitecture.md:
```diff
@@ -679,30 +696,123 @@ discovery.
+        If all available drivers are disabled using this environment variable,
+        then no drivers will be found by the loader and <b>will</b> result
+        in Vulkan applications failing to run properly.<br/>
+        This is also checked after other driver environment variables (such as
+        <i>VK_LOADER_DRIVERS_SELECT</i>) so it may disable drivers that were
+        intentionally enabled by that or other environment variables.
```
It is.  I missed this

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -134,11 +139,98 @@ If `VK_DRIVER_FILES` or `VK_ICD_FILENAMES` is present, then
-For security reasons, `VK_ICD_FILENAMES`, `VK_DRIVER_FILES` and
-`VK_ADD_DRIVER_FILES` are all ignored if running the Vulkan application with
-elevated privileges.
+There are new environment variables that will allow more control over what
+drivers the loader may use.
+These should also make it easier to diagnose driver problems or restrict usage
+to only a specific driver on a system.
```
👍🏽

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -134,11 +139,98 @@ If `VK_DRIVER_FILES` or `VK_ICD_FILENAMES` is present, then
+taking into account default search paths and other environment variables (like
+`VK_ICD_FILENAMES` or `VK_ADD_DRIVER_FILES`).
+
+When a driver is disabled using the new `VK_LOADER_DRIVERS_SELECT` filter, and
```
👍🏽

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -134,11 +139,98 @@ If `VK_DRIVER_FILES` or `VK_ICD_FILENAMES` is present, then
+These should also make it easier to diagnose driver problems or restrict usage
+to only a specific driver on a system.
+
+The new driver select environment variable `VK_LOADER_DRIVERS_SELECT` is a
```
👍🏽

MarkY-LunarG commented on docs/LoaderLayerInterface.md:
```diff
@@ -412,14 +419,136 @@ following:
+such as `VK_KHRONOS_LAYER_synchronization2` which could cause some applications
+to misbehave.
+
+When a layer is disabled using the new `VK_LOADER_LAYERS_DISABLE` filter, and
```
👍🏽

MarkY-LunarG commented on docs/LoaderInterfaceArchitecture.md:
```diff
@@ -750,6 +860,34 @@ may be removed in a future loader release.
+        <i>ppEnabledLayerNames</i>.
+    </small></td>
+    <td><small>
+        This has been replaced by <i>VK_LOADER_LAYERS_ENABLE</i>.
+        Similar to the new environment variable, if this is used it will also
+        override any layers disabled with <i>VK_LOADER_LAYERS_DISABLE</i>.
```
👍🏽

MarkY-LunarG commented on docs/LoaderInterfaceArchitecture.md:
```diff
@@ -679,30 +696,123 @@ discovery.
+        default search paths and other environment variables
+        (like <i>VK_LAYER_PATH</i>).
+        <br/>
+        This has replaced the older deprecated environment variable
+        <i>VK_INSTANCE_LAYERS</i>, however the older environment variable will
+        continue to work for some time.
```
👍🏽

MarkY-LunarG commented on docs/LoaderDebugging.md:
```diff
@@ -0,0 +1,312 @@
+
+### Selectively Re-enable Layers
+
+Each layer can be written down to know which layers were disabled.
+Then, if the problem goes away, each layer can then be individually re-added
+later to see if the problem re-appears.
```
👍🏽

MarkY-LunarG commented on docs/LoaderDebugging.md:
```diff
@@ -0,0 +1,312 @@
+  * What layers are found
+  * Where they were found
+  * If they are implicit what environment variables can be used to disable them
+  * If there is any incompatibility with a given layer
```
👍🏽

MarkY-LunarG commented on docs/LoaderDebugging.md:
```diff
@@ -0,0 +1,312 @@
+but this will provide more layer-specific info as well such as:
+  * What layers are found
+  * Where they were found
+  * If they are implicit what environment variables can be used to disable them
```
👍🏽

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -134,11 +139,98 @@ If `VK_DRIVER_FILES` or `VK_ICD_FILENAMES` is present, then
+taking into account default search paths and other environment variables
+(like `VK_ICD_FILENAMES` or `VK_ADD_DRIVER_FILES`).
+
+When a driver is disabled using the new `VK_LOADER_DRIVERS_DISABLE` filter, and
```
👍🏽

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -134,11 +139,98 @@ If `VK_DRIVER_FILES` or `VK_ICD_FILENAMES` is present, then
+
+#### Comma-delimited lists
+
+All these new filter environment variables accept comma-delimited input.
```
👍🏽

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -134,11 +139,98 @@ If `VK_DRIVER_FILES` or `VK_ICD_FILENAMES` is present, then
+This is especially important because it is difficult sometimes to determine the
+full name of a driver manifest file.
+So, instead of having to type in `intel_icd.x86_64.json` to select only the
+Intel driver on Linux, the substring `*intel*` can be used in the new
```
👍🏽

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -134,11 +139,98 @@ If `VK_DRIVER_FILES` or `VK_ICD_FILENAMES` is present, then
+
+#### Case-insensitive
+
+All these new filter environment variables assume the strings inside of the glob
```
👍🏽

MarkY-LunarG commented on docs/LoaderLayerInterface.md:
```diff
@@ -412,14 +419,136 @@ following:
+layer list in the loader for each application.
+These layers are enabled after implicit layers but before other explicit layers.
+
+When a layer is enabled using the new `VK_LOADER_LAYERS_ENABLE` filter, and
```
👍🏽

MarkY-LunarG commented on docs/LoaderLayerInterface.md:
````diff
@@ -412,14 +419,136 @@ following:
+WARNING | LAYER:  Layer "VK_LAYER_LUNARG_wrap_objects" force enabled due to env var 'VK_LOADER_LAYERS_ENABLE'
+```
+
+The new layer disable environment variable `VK_LOADER_LAYERS_DISABLE` is a
````
👍🏽

MarkY-LunarG commented on docs/LoaderLayerInterface.md:
```diff
@@ -412,14 +419,136 @@ following:
+
+#### Comma-delimited lists
+
+All these new filter environment variables accept comma-delimited input.
```
👍🏽

MarkY-LunarG commented on docs/LoaderLayerInterface.md:
```diff
@@ -412,14 +419,136 @@ following:
+full name of a driver manifest file.
+So, instead of having to type in `VK_LAYER_KHRONOS_validation` into the older
+`VK_INSTANCE_LAYERS` environment variable to force on validation, the substring
+`*validation` can be used in the new `VK_LOADER_LAYER_ENABLE` environment
```
👍🏽

MarkY-LunarG commented on docs/LoaderLayerInterface.md:
```diff
@@ -412,14 +419,136 @@ following:
+
+#### Case-insensitive
+
+All these new filter environment variables assume the strings inside of the glob
```
👍🏽

MarkY-LunarG commented on docs/LoaderLayerInterface.md:
```diff
@@ -412,14 +419,136 @@ following:
 
+### Layer Filtering
+
+There are new environment variables that will allow more control over what
+layers the loader may load.
+These should also make it easier to diagnose layer problems or restrict usage
+of any layers that may cause problems.
```
👍🏽

MarkY-LunarG commented on loader/vk_loader_platform.h:
```diff
@@ -86,11 +86,22 @@
 // Environment Variable information
-#define VK_ICD_FILENAMES_ENV_VAR "VK_ICD_FILENAMES"  // Deprecated
+#define VK_ICD_FILENAMES_ENV_VAR "VK_ICD_FILENAMES"  // Deprecated in v1.3.207 loader
 #define VK_DRIVER_FILES_ENV_VAR "VK_DRIVER_FILES"
-#define VK_ADDITIONAL_DRIVER_FILES_ENV_VAR "VK_ADD_DRIVER_FILES"
 #define VK_LAYER_PATH_ENV_VAR "VK_LAYER_PATH"
+// Support added in v1.3.207 loader
+#define VK_ADDITIONAL_DRIVER_FILES_ENV_VAR "VK_ADD_DRIVER_FILES"
 #define VK_ADDITIONAL_LAYER_PATH_ENV_VAR "VK_ADD_LAYER_PATH"
+// Support added in v1.3.217 loader
```
I'll leave this comment open until then.

charles-lunarg replied:
I will add that this version should be added to the docs, to indicate that it only is available in X version and onwards. 

> `VK_LOADER_LAYERS_ENABLE` is available in all loaders version 1.3.YYY and above. 

Would be good to add this to the env-var table as well. Like we've added several env-vars now, best to make sure people know what loader versions can use them.

MarkY-LunarG replied:
I've added the 1.3.yyyy note here and in several other places including the env var table.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 24309.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1483 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1483 failed.

MarkY-LunarG commented on loader/loader.c:
```diff
@@ -767,10 +768,241 @@ VkResult loader_add_to_dev_ext_list(const struct loader_instance *inst, struct l
-bool loader_add_meta_layer(const struct loader_instance *inst, const struct loader_layer_properties *prop,
-                           struct loader_layer_list *target_list, struct loader_layer_list *expanded_target_list,
-                           const struct loader_layer_list *source_list);
+// Determine the type of filter string based on the contents of it.
```
It's only 400 total lines.  Do you really find it that intrusive in the main file here?

MarkY-LunarG replied:
I moved it and a few other functions into get_environment.h/.c and I renamed those files loader_environment.h/.c

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 24328.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1484 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 24359.

MarkY-LunarG commented:
Rebased on latest master due to conflicts.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1485 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1485 failed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 24396.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1488 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1488 failed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 24687.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1489 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1489 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 24751.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1491 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1491 passed.

MarkY-LunarG commented:
@hadess, if you see anywhere in the updated docs this is still the case, please let me know.  I tried to sanitize it in my latest round but may have missed some instances.  Thnx

charles-lunarg commented on tests/loader_layer_tests.cpp:
```diff
@@ -602,16 +1253,8 @@ TEST(OverrideMetaLayer, InvalidDisableEnvironment) {
-
-    InstWrapper inst{env.vulkan_functions};
-    inst.CheckCreate();
+    EXPECT_EQ(VK_ERROR_INITIALIZATION_FAILED, env.vulkan_functions.vkEnumerateInstanceLayerProperties(&layer_count, nullptr));
+    EXPECT_EQ(layer_count, 0U);
```
This is fishy to me, since the explicit layer should be found normally, since its in the standard layer search paths.

MarkY-LunarG replied:
Ah, I caught a real bug.  I thought my code removed all layers appropriately but it was incorrect as you pointed out.  fixing.

charles-lunarg commented on loader/loader.c:
```diff
@@ -2532,6 +2570,10 @@ static VkResult loader_read_layer_json(const struct loader_instance *inst, struc
     if (VK_SUCCESS != result && NULL != props) {
         // Make sure to free anything that was allocated
         loader_remove_layer_in_list(inst, layer_instance_list, props_index);
+        if (layer_instance_list->count == 0) {
+            // Cleanup any allocated memory
+            loader_delete_layer_list_and_properties(inst, layer_instance_list);
+        }
```
Why are we removing the *entire* layer list if only one of the entries has a non-success return value? Shouldn't it only do this for catastrophic errors?

I worry about the case where a layer is 'invalid' and it nukes the rest of  the list.

MarkY-LunarG replied:
I think it was related to my bad code for the above.  Undoing.

charles-lunarg commented on docs/LoaderLayerInterface.md:
```diff
@@ -412,14 +419,136 @@ following:
+the layer is Implicit or Explicit).
+This means that they will not be added to the enabled layer list in the loader
+for each application.
+This could mean that layers requested by an application are also not enabled
+such as `VK_KHRONOS_LAYER_synchronization2` which could cause some applications
+to misbehave.
```
If an application asks for layer X to be enabled in instance creation, but VK_LOADER_LAYERS_DISABLE is set to disable that layer, wouldn't the app get VK_ERROR_LAYER_NOT_PRESENT in instance creation? That would be my understanding. 

Put another way, does vkEnumerateInstanceLayerProperties include disabled layers? If not, then the app shouldn't 'try' to enable a layer that doesn't exist according to vkEnumerateInstanceLayerProperties.

MarkY-LunarG replied:
So vkEnumerateInstanceLayerProperties no longer shows that layer.  So, if they are looking for the layer, it won't be there.  If they blindly enable it, vkCreateInstance ignores that layer inside the driver and only reports messages via it's log.

charles-lunarg replied:
Doing a bit of playing around, it seems that `VK_LOADER_LAYERS_DISABLE` disabling a layer that the application is trying to enable in `VkInstanceCreateInfo` does *not* provoke `vkCreateInstance` returning an error. It disables the layer and lets the app create an instance.

However, when calling `vkEnumerateInstanceLayerProperties()` the layer is still present in the list. I think this will cause confusion because the layers apps try to enable will appear when they look for it but aren't enableable.

charles-lunarg replied:
The behavior that was settled on was to fail instance creation if an application tries to enable a 'disabled' layer with VK_ERROR_LAYER_NOT_PRESENT, as well as 'filter' out disabled layers from `vkEnumerateInstanceLayerProperties()`.

MarkY-LunarG commented on loader/loader_environment.c:
```diff
@@ -0,0 +1,546 @@
+/*
+ *
+ * Copyright (c) 2014-2021 The Khronos Group Inc.
+ * Copyright (c) 2014-2021 Valve Corporation
+ * Copyright (c) 2014-2021 LunarG, Inc.
```
👍🏽

hadess commented on docs/LoaderDebugging.md:
```diff
@@ -0,0 +1,322 @@
+
+You can now use the filtering environment variables
+(`VK_LOADER_DRIVERS_SELECT` and `VK_LOADER_DRIVERS_DISABLE`) to selectively
+select or disable various drivers.
```
`selectively enable or disable`, rather than `selectively select`?

MarkY-LunarG replied:
No, the drivers are enabled by default.  The process for drivers is select (i.e. force a particular driver) and disable.

MarkY-LunarG replied:
Ah good point...  How about:

```
You can now use the filtering environment variables
(`VK_LOADER_DRIVERS_SELECT` and `VK_LOADER_DRIVERS_DISABLE`) to control what
drivers the loader discovers.
```

hadess replied:
```
You can now use the filtering environment variables
(`VK_LOADER_DRIVERS_SELECT` and `VK_LOADER_DRIVERS_DISABLE`) to control what
drivers the loader will attempt to load.
```
maybe? Otherwise your version is fine.

hadess commented on docs/LoaderDebugging.md:
```diff
@@ -0,0 +1,322 @@
+You can now use the filtering environment variables
+(`VK_LOADER_DRIVERS_SELECT` and `VK_LOADER_DRIVERS_DISABLE`) to selectively
+select or disable various drivers.
+Remember, that to disable drivers, you must use the name of the Driver JSON to
```
This whole sentence needs reworking:
`to disable drivers, you must use the name of the Driver JSON to disable since...`

MarkY-LunarG replied:
How about this instead?

```
For drivers, the string globs passed into the above environment variables will
be compared against the driver JSON file name since there is no driver name
known to the loader until much later in the Vulkan initialization process.
```

hadess replied:
Looks good.

hadess commented on docs/LoaderDebugging.md:
````diff
@@ -0,0 +1,322 @@
+```
+
+These can be used to make sure that the appropriate drivers are enabled/disabled
+properly.
````
`properly`? What would improperly enabled/disabled drivers be? ;)

I would just remove the sentence.

MarkY-LunarG replied:
👍🏽

hadess reviewed (commented):
A few things I found. Hope this helps.

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -126,19 +131,103 @@ There may be times that a developer wishes to force the loader to use a specific
+The driver disable environment variable `VK_LOADER_DRIVERS_DISABLE` is a
+comma-delimited list of globs to search for in known drivers.
+Since drivers don’t have a name like layers, this substring is used to compare
+against the manifest filename.
```
👍🏽

MarkY-LunarG commented on docs/LoaderDriverInterface.md:
```diff
@@ -126,19 +131,103 @@ There may be times that a developer wishes to force the loader to use a specific
+#### Environment Variable Priority
+
+The values from the disable environment variable will be considered
+<b>before</b> the select environment variable.
```
👍🏽

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 27262.

MarkY-LunarG commented on docs/LoaderLayerInterface.md:
```diff
@@ -412,14 +419,130 @@ following:
+`~all~` or `~explicit~` could cause application breakage if the application is
+relying on features provided by one or more explicit layers.
+
+#### Comma-delimited lists
```
Moved this duplicate section into LoaderINterfaceArchitecture.md and referenced it in both locations.

MarkY-LunarG commented:
@charles-lunarg:

> Put another way, does vkEnumerateInstanceLayerProperties include disabled layers? If not, then the app shouldn't 'try' to enable a layer that doesn't exist according to vkEnumerateInstanceLayerProperties.

It did not, I fixed this with the latest changes.

> I will add that this version should be added to the docs, to indicate that it only is available in X version and onwards.
>
> > VK_LOADER_LAYERS_ENABLE is available in all loaders version 1.3.YYY and above.
>
> Would be good to add this to the env-var table as well. Like we've added several env-vars now, best to make sure people know what loader versions can use them.

I've placed 1.3.yyyy everywhere this will need to be updated.  Once passing and approved, I'll update before the push.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1496 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1496 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 27646.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1497 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1497 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 27921.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1502 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1502 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 7460.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1522 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1522 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 23923.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1585 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1585 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 24177.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1589 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1589 passed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 26475.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1601 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1601 failed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 27213.

charles-lunarg reviewed (approved):
After much finagling, this is in a good shape to be merged.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1604 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1604 failed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 27229.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1605 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1605 failed.

ci-tester-lunarg commented:
CI Vulkan-Loader build queued with queue ID 27266.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1606 running.

ci-tester-lunarg commented:
CI Vulkan-Loader build # 1606 passed.
